### PR TITLE
priorities: CONNECT can be prioritized too

### LINKED ISSUE #1495: The priority extension doesn't mention the CONNECT method
The CONNECT method is somewhat special in the sense that most of the data transmitted on the stream is not a request or response. If a server receives a PRIORITY_UPDATE frame that references an open client-initiated bidirectional stream on which a CONNECT request was previously received, what is the server supposed to do?

This line of reasoning came thanks to @ianswett and @LPardue on https://github.com/ietf-wg-masque/draft-ietf-masque-h3-datagram/issues/46

### PR DESCRIPTION
**Pull request by LPardue**
fixes #1495

### CODE REVIEW

**Comment by LPardue** [comment]
@DavidSchinazi PTAL

**Review by DavidSchinazi — APPROVED** [review]
Thanks for writing this!

**Review by kazuho — APPROVED** [review]
Yes, this is something worth documenting. Thank you for the PR.

File: draft-ietf-httpbis-priority.md [hunk 1]
```diff
@@ -631,6 +631,12 @@ suboptimal. How ever a server decides to schedule a pushed response, it can
 signal the intended priority to the client by including the Priority field in
 a PUSH_PROMISE or HEADERS frame.
 
+# Scheduling and the CONNECT Method
+
+When a request stream carries the CONNECT method, the scheduling guidance in
+this document applies to the DATA frames that carry either HTTP content or
+payload related to the established TCP connection.
```
Comment (tfpauly) [inline draft-ietf-httpbis-priority.md:638]:
```suggestion
payload related to the established TCP connection. Streams that use CONNECT often
are handled as unstructured streams of bytes; for such cases, clients can mark the
incremental parameter. Servers MAY treat CONNECT streams as incremental even
if the client does not indicate the parameter.
```

Comment (tfpauly) [reply draft-ietf-httpbis-priority.md:638]: @LPardue  thoughts?

Comment (LPardue) [reply draft-ietf-httpbis-priority.md:638]: thanks, this helps.

I think this is actually a good example of a danger we point out in the Server scheduling section

> At the same urgency level, an incremental request of indeterminate length followed by a non-incremental large resource.

So how about tweaking your suggestion to say something like 

"When a request stream carries the CONNECT method, the scheduling guidance in this document applies to all DATA frames on the stream. If the request is unsuccesful, there might be DATA frames that carry HTTP content. If the request is successful, DATA frames carry payload related to a TCP connection. This is often handled as an unstructured byte stream of indeterminate length, where chunks are processed incrementally. Clients can send CONNECT requests with the incremental parameter. Servers can use the incremental parameter, the CONNECT method itself, the response status, and other factors as input into their scheduling; see {{server-scheduling}}."




    
    
      
    this document applies to the DATA frames that carry either HTTP content or




    
    
      
    <span class="blob-code-inner blob-code-marker" data-code-marker="+"><span class="pl-s">payload related to the established TCP connection.</span></span>

Comment (martinthomson) [reply draft-ietf-httpbis-priority.md:638]: Maybe just "all frames on the stream", not making DATA special in any way.  Then you don't need to talk about the protocol bits (which other documents cover).

Regarding incremental, I would just give permission to treat CONNECT as incremental even if the priority didn't indicate that it was.  Nothing else makes sense really.

Comment (kazuho) [reply draft-ietf-httpbis-priority.md:638]: > Regarding incremental, I would just give permission to treat CONNECT as incremental even if the priority didn't indicate that it was. Nothing else makes sense really.

I'm not entirely sure. A client could do interesting things like create one CONNECT tunnel for each priority level, and when it is about to send a HTTP request, choose one of the tunnel based on the request priority.

Comment (martinthomson) [reply draft-ietf-httpbis-priority.md:638]: The tunnel doesn't know that you are using this for HTTP requests.

The problem I have with non-incremental is that you can't use the CONNECT tunnel for anything.  It's fundamentally TCP, which requires that you have a continuous exchange of bytes.  There is no sense in saying non-incremental because while that implies a very particular exchange pattern: this stream is not useful to me unless I have all of it.  

As soon as you have TLS inside a tunnel, non-incremental is non-viable.

I can see value in having priority on different tunnels, that's a good way of (somewhat) dealing with remote nodes who aren't able to do proper prioritization.  But non-incremental: that implies a very particular interaction model.

Comment (kazuho) [reply draft-ietf-httpbis-priority.md:638]: I think we agree that some sort of bandwidth sharing is necessary for CONNECT requests, as TCP connections often have idle timeouts. The question is if incremental is sufficient and adequate.

Regarding sufficiency, there is at least one corner case. Consider the case where one H3 stream is used to carry GET (u=0) and CONNECT (u=3,i). GET (u=0) would absorb all available bandwidth until that response is complete. This is not good, therefore, we need to give some more advice than "assume incremental."

Regarding adequacy, incremental assumes a particular way of bandwidth sharing; i.e. treat all streams equal. As stated above, that might not always what we'd want.

Taking a step back, idle timeout is an issue even for intermediaries forwarding HTTP messages. Let's say that the intermediary is connected to upstream using HTTP/1.1. If that intermediary does not read more data from upstream, the upstream might consider that the HTTP/1.1 connection has stalled, and close it. To prevent that, the intermediary has to give a small amount of bandwidth to that message.

To summarize, I think the advice that we might want to give are:
* servers SHOULD give some bandwidth to streams when idle timeout is a concern (regardless of they carry HTTP messages or CONNECT data)
* CONNECT clients can set the incremental flag if it wants the server to provide fairness between the CONNECT requests that the client issues.

**Comment by LPardue** [comment]
This sounds reasonable. You first bullet could go into the server scheduling section. The second bullet could go into CONNECT specific section added by this PR, probably mingled with Tommy's suggestion.

File: draft-ietf-httpbis-priority.md (hunk 1 above)
Comment (kazuho) [reply draft-ietf-httpbis-priority.md:638]: Regarding the first bullet point, I've opened https://github.com/httpwg/http-extensions/pull/1562. I would appreciate your feedback. Thank you in advance.

**Review by kazuho — APPROVED** [review]

**Comment by LPardue** [comment]
Via https://github.com/httpwg/http-extensions/pull/1562 and some further tweaks to the text, I think we've addressed the feedback. Merging this now.
